### DISCUSSION TITLE
I've never seen such a confusing sample app

### DISCUSSION THREAD

**Discussion by csbenz** [patrykandpatrick/vico#527] [Ideas]
Am I the only one here? Why isn't there a simple example included? For example, I want stacked column bars in my graph. I've been searching the documentation and the sample app since 2 hours and still have no idea how to do it.
Additionally, when using views, all objects to set attributes have to be cast? What is this about?

Perhaps we could include a simple and basic example in the sample app (no android architecture or frameworks, just code that shows how the library works and operates).

**Comment by csbenz** [comment]
I've understood it's because this library is mainly meant for Compose. Is there available samples and examples for use with views?

**Comment by Gowsky** [comment]
Hi @csbenz, no, Vico is not "mainly meant for Compose". However, the sample app is built with Jetpack Compose for convenience.
The sample app mostly consists of the [ShowcaseScreen](https://github.com/patrykandpatrick/vico/blob/89d0a098126a33227a98c86c39d37ed186678208/sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/ShowcaseScreen.kt), [ShowcaseViewModel](https://github.com/patrykandpatrick/vico/blob/89d0a098126a33227a98c86c39d37ed186678208/sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/ShowcaseViewModel.kt) and [charts](https://github.com/patrykandpatrick/vico/tree/89d0a098126a33227a98c86c39d37ed186678208/sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/charts) (both composable and view variants) and we find it as simple as it can be.

**Comment by csbenz** [comment]
This library is driving me insane, am I the only one?

**Reply by patrickmichalik** [reply]
We’re sorry to hear that. Vico’s API is quite large. While this contributes to the library’s flexibility, it can also pose a bit of a challenge to new users. Every API is documented and listed in the API reference, which we encourage people to check out. We’re working on simplifications to make it easier to get started with Vico. Please feel free to share your suggestions and reach out if you’re unsure about anything.

**Comment by patrickmichalik** [comment]
Hello! I wanted to provide some extra context. While the sample app is built in Jetpack Compose, each chart included in it has a Compose-based variant and a view-based variant. Standard XML layouts and `AndroidViewBinding` are used for the view-based charts. What you see in the `AndroidViewBinding` lambdas is what you could put in, for example, `Activity#onCreate`. Vico is compatible with both UI systems—there are no Compose-specific or view-specific features. In the documentation, Compose and the view system are considered equally. The sample app has a relatively standard structure, but several simplifications are being worked on. Please feel free to share any suggestions you may have.

Regarding stacked columns, the fifth sample chart can serve as an example. [Here](https://github.com/patrykandpatrick/vico/blob/master/sample/src/main/res/values/chart_5_styles.xml#L63) is the relevant line of `chart_5_styles.xml`. (For Vico 1, please see [here](https://github.com/patrykandpatrick/vico/blob/v1/sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/charts/Chart5.kt#L91).)

About the need for casting, it arises when you instantiate classes indirectly by using XML attributes. This comes down to how typing works in general. If you programmatically create a `CartesianChart` with a `ColumnCartesianLayer`, no casting is needed. If you use XML attributes, there’s no way for the typing system to know that `cartesianChartView.chart?.layers[0]` is a `ColumnCartesianLayer`, as opposed to any other `CartesianLayer`, and so a cast to `ColumnCartesianLayer` is needed. There’s nothing Vico can do about this. However, XML is optional—if you prefer, you can use Vico entirely programmatically and avoid casting.

**Reply by Randomboiii03** [reply]
Hello, could you please make some more Compose examples and tutorials for beginners? The current ones are a bit tough to chew. Thanks much appreciated.

**Reply by patrickmichalik** [reply]
Sure thing, @Randomboiii03. This is something that we’re currently looking into (for both Compose and views). If you have any specific requests, please let me know.

**Reply by Randomboiii03** [reply]
Clearer documentation with examples or a video tutorial would be incredibly helpful for creating charts. The current resources make it unnecessarily difficult for beginners. Thank you very much 😁

**Reply by patrickmichalik** [reply]
Thanks, @Randomboiii03! We’ll keep that in mind.

**Reply by Dymoxz** [reply]
Hi, is there any progress on clearer examples or tutorials? Im confused about how to create a simple graph lol

**Reply by virtualpathum** [reply]
I agree. it is not easy to do simple things. For eg: I just wanted to show simple 0,5,10,15 .. values for my y-axis and still couldn't find how to do that.. although it is a robust library it is tough to understand how to do simple things.

**Reply by patrickmichalik** [reply]
@Dymoxz, I presume you’re referring to Vico 2. Please let me know if not. Improvements are being made across the sample app and the wiki. See [here](https://github.com/patrykandpatrick/vico/blob/master/sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/charts/Chart1.kt) for the new version of the first sample chart, for example. The data is created right in `Chart1`, the line color is customized more directly, and other simplifications have been made. Regarding creating a simple chart, the wiki now includes examples of [a basic column chart](https://patrykandpatrick.com/vico/wiki/2.0.0-alpha.8/cartesian-charts/layers/column-layer#basic) and [a basic line chart](https://patrykandpatrick.com/vico/wiki/2.0.0-alpha.8/cartesian-charts/layers/line-layer#basic), which I believe should help. If they don’t, we’d appreciate it if you let us know what you’re having trouble with.

@virtualpathum, the APIs needed for this are covered in the wiki, but we’ll add a dedicated paragraph for this use case. We also plan on introducing an API to make _y_-axis step customization easier, and we’ll showcase this API in the sample app. In the meantime, please feel free to open a discussion about this in the “Q&A” category, and we’ll get back to you.

Thanks for the feedback!

**Comment by zell-mbc** [comment]
Maybe some additional feedback from someone who heard of Vico for the first time 2 days ago and now has got his first chart converted over.
At first I was pretty confused when looking at the sample code as well, but I managed.
What would have helped me I think would have been some comments in the code, explaining what things do and why you need them.
E.g.:
// The modelProducer is the entity which…well, I don't know (yet) what it does other than taking lists of columns
// It consumes any number of lists, datatypes should be numeric, if you need strings for e.g. the x axis you convert as show in the bottomAxisValueFormatter example…
 modelProducer.tryRunTransaction { columnSeries { series(List(47) { 2 + Random.nextFloat() * 18 }) } }

The second thing which still confuses me is that  sample data generation is somehow obfuscating what the Vico functions need as input and what's there to create sample data. 
Simple lists with ~10 static values as input would make the initial steps easier.

**Reply by patrickmichalik** [reply]
Hello, and thanks for the feedback! Regarding the first point, could you elaborate on how this would tie into the broader picture, which includes the wiki and the API reference? The comments in your example appear to match what’s found on the [“Data,”](https://patrykandpatrick.com/vico/wiki/cartesian-charts/data) [“Column layer,”](https://patrykandpatrick.com/vico/wiki/cartesian-charts/layers/column-layer) and [“Axes”](https://patrykandpatrick.com/vico/wiki/cartesian-charts/axes) wiki pages. It’s my impression that the wiki and the API reference are better suited for such general documentation than the sample app is. As it stands, the sample app is meant to complement the other two resources. We recommend starting by reading the wiki, where you can find an overview of how Vico works. However, we’re open to suggestions for how things could be reorganized.

As for the second point, I see how the data generation logic can be a bit confusing. Simpler examples can be found in the wiki—see, for example, the [“Data”](https://patrykandpatrick.com/vico/wiki/cartesian-charts/layers/column-layer/#data) section of the [“Column layer”](https://patrykandpatrick.com/vico/wiki/cartesian-charts/layers/column-layer) page. That said, we’ll look into making some of the sample charts use static data, as you’ve suggested, to make things easier to understand.

**Reply by zell-mbc** [reply]
Some of this information would probably be identical, and I guess different people approach things differently.
I had a brief look at the Wiki and API initially and figured that by the look of it everything I need seems to be to be there, very high level review though.

My approach to new things like your tool is to start with something as minimalistic as possible and then (with the help of whatever documentation is available) learn and add the bells and whistles one by one. 
So at first I want to get something visible on the screen to iterate over, hence I quickly moved over to my app and your demo code.
Starting with the simplest chart1, when looking at the sample code it felt like there's a lot going on and it took me a while to separate demo data from the actual library. Until I realized that it's really just a bunch of lists<Number> I need to create and that's all.
 
I spent a good amount of time this afternoon with the CandleChart demo code, looking at the RandomCartesianModelGenerator I was like, sh*t there's a hell lot going on to feed the chart. A lot more complicated than the list chart I set up yesterday. Well, a few hours later I realized this is actually the exact same thing. Feed the chart with 4 or 5 lists and that's it…all this highly sophisticated code really is just to create these 5 lists with good looking data.
After I realized that, feeding the chart with my existing data was dead easy, a one-liner because I had those lists already:
                            `add(CandlestickCartesianLayerModel.partial(timeline, dia, sys, dia, sys))`


I understand that you want your demo to look cool, and it does, but maybe a way around this is to have the simplest implementation possible as a comment, next to where you call your random generators?
 ```
// You need to add 5 (4) lists with identical size to the add function (x-axis, opening, closing,…
// Passing the lists through CandlestickCartesianLayerModel.partial converts the lists into objects suitable for the chart.
// val xAxis = listOf<Int>(1,2,3)
// val opening = listOf<Int>(10,12,9)
// …
// add(CandlestickCartesianLayerModel.partial(xAxis, opening,…)
```
I am sure you got this covered in the Wiki somewhere. And now that I look at it I feel like I should have figured this out a lot lot quicker, but probably my brain got lost in all the (cool) code.

By the way, let me say that I am really impressed by Vico, never mind my comments on the demo app, I still was able to get something meaningful up and running in no time, and this stuff looks really good! So thank you very much for Vico and all the work you put in! As far as I am concerned I will put it to good use in my app helping people with health issues.

**Reply by patrickmichalik** [reply]
Thanks for the insights and the kind words, @zell-mbc! We’re glad to hear that you’ve found Vico useful and that it’s being used in such a meaningful project.

There are starter examples with static data in the wiki. For instance, you can find a basic `LineCartesianLayer` implementation to iterate on [here](https://patrykandpatrick.com/vico/wiki/cartesian-charts/layers/line-layer#basic). There’s information on how to handle `LineCartesianLayer` data [here](https://patrykandpatrick.com/vico/wiki/cartesian-charts/layers/line-layer#data). Finally, [here](https://patrykandpatrick.com/vico/wiki/cartesian-charts/layers/candlestick-layer#data) is the equivalent section for `CandlestickCartesianLayer`. You’ll notice that the code snippets in this section closely match the snippet in the example comment you’ve shared. I believe the confusion described would’ve been avoided if these parts of the wiki had been referred to, but I understand that the sample app is where many people go first. 

Since the wiki appears to include the information in question (which isn’t to say it won’t be expanded) and is preferable to code comments as a documentation environment, we could add comments with relevant wiki links to the sample app’s code. This would provide the same benefits as the suggested approach, in effect, without requiring us to introduce a third documentation location and duplicate existing documentation in a restrictive format. We’d introduce such link comments in addition to expanding the sample app with simpler examples, as mentioned above. What are your thoughts on such a solution? 

We’ll also be looking into improving the KDoc documentation, which is accessible both from external projects and from the sample app’s code. Many of the descriptions therein can be made more practical and complemented with simple examples (like the ones you’ve shared and those in the wiki). This will slightly increase the overlap between the wiki and the API reference, but this overlap is unavoidable either way. That said, we can also add wiki links to the API reference—the optimal balance will have to be determined.

I wanted to briefly refer back to `CandlestickCartesianLayer`. First, I wanted to add that it was introduced just over a week ago, and more documentation and examples are on the way. Second, please note that the recommended means of adding `CandlestickCartesianLayer` data to a `CartesianChartModelProducer.Transaction` is `candlestickSeries`, as described in the last wiki section linked above. This is a bit more concise than the combination of `CandlestickCartesianLayerModel.partial` and `CartesianChartModelProducer.Transaction#add`. `CandlestickCartesianLayerModel.partial` is used in `RandomCartesianModelGenerator` because there’s no `CartesianChartModelProducer.Transaction` there.

**Reply by zell-mbc** [reply]
You are right, it's all there, including the exact simple examples I have been proposing. In hindsight I would have been better off by focusing on the Wiki instead of the sample app. I also got sidetracked by the Extras concept described here https://patrykandpatrick.com/vico/wiki/cartesian-charts/axes which is new to me., so I thought let's see how this plays out in the demo app.
Probably another case of being unable to see the wood for the trees.
 
But then again, one has to start somewhere and you only know what would have been more effective after being through with it. And like stated earlier, I am still pleased with the outcome, I got to results significantly quicker than with my previous library.
I mean, there is a learning curve with everything new.

As far as the CandleStick chart goes, I'll check out  ```candlestickSeries``` , but after all I am only testing out the candle stick chart to see if I can shoehorn it into what I think fits better, a band chart.
Main use case is for capturing blood pressure values, where you always have an upper and lower value (open, close=systolic, diastolic), what I am envisioning is two line charts with the chart colour in the middle, but I don't think that's possible?

![image](https://github.com/patrykandpatrick/vico/assets/29018767/2fb7b1e8-0136-49fc-a778-2bb2a476d61c)

I will test this with my 10k records sample database soon…

By the way, is there a way to move the labels to the right? I tried helperline->rememberTextComponenet.>textAlignment = Layout.Alignment.ALIGN_OPPOSITE, but that didn't do the trick.

**Reply by Gowsky** [reply]
>what I am envisioning is two line charts with the chart colour in the middle, but I don't think that's possible?

Unfortunately, it's not possible currently, as mentioned in #893.

>By the way, is there a way to move the labels to the right? I tried helperline->rememberTextComponenet.>textAlignment = Layout.Alignment.ALIGN_OPPOSITE, but that didn't do the trick.

`HorizontalLine` allows you to change the label alignment with `horizontalLabelPosition` parameter. `TextComponent.textAlignment` controls the relative alignment of multiple lines of text with different lengths.

**Comment by Gowsky** [comment]
All of the updates to the sample app mentioned by @patrickmichalik are complete. This includes global simplifications to make it easier to follow and three new starter charts intended for beginners.

We've also made multiple other enhancements. This includes the additional improvements @patrickmichalik described, except for the KDoc expansion, which hasn't been fully looked into yet. Since the KDoc expansion is ongoing and not a key point in the context of this thread, and everyting else is done, I'm closing the discussion.

If you have any further suggestions, feel free to open a discussion in "Ideas."

I apologize if my previous comment came accross as harsh, but please ensure that your feedback is constructive. Comments like "I've never seen such a confusing sample app" or "This library is driving me insane" only add negativity, without contributing to any meaningful improvement. They don't point to any specific issues that can be addressed. Please also remember that this is a free and open-source project.

**Comment by csbenz** [comment]
nearly two years later and the documentation is even more confusing. this is a joke project documentation lol

**Reply by patrickmichalik** [reply]
Once again, we’d ask you to keep feedback constructive. Such disparaging drive-by comments are of no utility and only discourage contributors. Please engage civilly, or we’ll have to restrict your access to the repository.
